src/Microsoft.Azure.WebJobs.Host/Hosting/DrainModeManager.cs

```diff
@@ -0,0 +1,94 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using Microsoft.Azure.WebJobs.Host.Listeners;
+using Microsoft.Azure.WebJobs.Host.Protocols;
+using Microsoft.Azure.WebJobs.Logging;
+using Microsoft.Extensions.Logging;
+using System;
+using System.Collections;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Azure.WebJobs.Host
+{
+    internal class DrainModeManager : IDrainModeManager
+    {
+        private readonly ILoggerFactory _loggerFactory;
+        private readonly ILogger _logger;
+        private ConcurrentDictionary<Guid, CancellationTokenSource> _cancellationTokenSources;
+
+        public DrainModeManager(ILoggerFactory loggerFactory = null)
+        {
+            _loggerFactory = loggerFactory;
+            _logger = _loggerFactory.CreateLogger<DrainModeManager>();
+            _cancellationTokenSources = new ConcurrentDictionary<Guid, CancellationTokenSource>();
+        }
+
+        public bool IsDrainModeEnabled { get; private set; } = false;
+
+        public ICollection<IListener> Listeners { get; private set; } = new Collection<IListener>();
+
+        public void RegisterListener(IListener listener)
+        {
+            Listeners.Add(listener);
+        }
+
+        public void RegisterTokenSource(Guid guid, CancellationTokenSource tokenSource)
+        {
+            _cancellationTokenSources.TryAdd(guid, tokenSource);
+        }
+
+        public void UnRegisterTokenSource(Guid guid)
+        {
+            _cancellationTokenSources.RemoveIfContainsKey(guid);
+        }
+
+        public async Task EnableDrainModeAsync(CancellationToken cancellationToken)
+        {
+            if (!IsDrainModeEnabled)
+            {
+                IsDrainModeEnabled = true;
+                _logger.LogInformation($"DrainMode is set to {IsDrainModeEnabled}");
```

Comment:
Won't this always be true? We don't need the interpolation here. Changing the message to `Drain mode enabled.` is a better approach.